pyrokinetics/kinetics/KineticsReaderpFile.py

```diff
@@ -0,0 +1,241 @@
+"""
+Reads in an Osborne pFile: https://omfit.io/_modules/omfit_classes/omfit_osborne.html#OMFITosborneProfile
+Note that pFile units are: density 10^20 m^-3 temp keV
+For Pyrokinetics, we need these in m^-3 and eV, respectively.
+"""
+
+import re
+from textwrap import dedent
+from typing import Dict, Optional
+
+from .KineticsReader import KineticsReader
+from .pfile_reader import PFileReader
+from ..constants import electron_mass, hydrogen_mass
+from ..equilibrium.equilibrium import read_equilibrium
+from ..species import Species
+from ..typing import PathLike
+
+import numpy as np
+from scipy.interpolate import InterpolatedUnivariateSpline
+
+
+def ion_species_selector(nucleons, charge):
+    """
+    Returns ion species type from:
+
+    hydrogen deuterium, tritium, helium3, helium, other impurity.
+
+    Might need to update with more specific masses, such as 6.94 for Li-7, etc.
+    """
+    if nucleons == 1:
+        if charge == 1:
+            return "hydrogen"
+        else:
+            print(
+                "You have a species with a single nucleon which is not a proton. Strange. Returning neutron for now. \n"
+            )
+            return "neutron"
+    if nucleons == 2 and charge == 1:
+        return "deuterium"
+    elif nucleons == 4 and charge == 2:
+        return "helium"
+    elif nucleons == 3:
+        if charge == 1:
+            return "tritium"
+        if charge == 2:
+            return "helium3"
+    else:
+        return "impurity"
+
+
+def np_to_T(n, p):
+    """
+    n is in m^{-3}, T is in eV, p is in Pascals.
+    Returns temperature in eV.
+    """
+    return np.divide(p, n) / ((1.381e-23) * 11600)
+
+
+class KineticsReaderpFile(KineticsReader):
+    def read(
+        self,
+        filename: PathLike,
+        eq_file: PathLike = None,
+        time_index: int = -1,
+        time: Optional[float] = None,
+    ) -> Dict[str, Species]:
+        """
+        Reads in Osborne pFile. Your pFile should just be called, pFile.
+        Also reads a geqdsk file via eq_file to obtain r/a.
+        """
+        # eq_file must be provided
+        if eq_file is None:
+            raise ValueError(
+                dedent(
+                    f"""\
+                    {self.__class__.__name__} must be provided with a G-EQDSK file via
+                    the keyword argument 'eq_file'.
+                    """
+                )
+            )
+
+        # Read pFile, get generic data.
+        pFile = PFileReader(str(filename))
+        psi_n = pFile.__getattribute__("ne").x
```

Comment:
Isn't it possible in a pFile for different quantities to have different psinorm grids?
In this method I think we're assuming all quantities live on the save psinorm grid as given by the electron density values?

Reply:
@d7919 @bpatel2107  Yes David, you're completely correct. It's not only possible, but normal for different quantities to have different psinorm grids.